### PR TITLE
Letter api step 1 - refactor

### PR DESCRIPTION
**Pull request by leohemsted**
In the lead-up to sending notifications via API, here's a PR that refactors the v2 post_notifications functions to get ready for the letters implementation. The actual processing of the letter job isn't in place yet, but this PR does some refactoring of the post_notification function.

It also adds the letter schema and currently skipped tests for the letter endpoint

### CODE REVIEW

File: tests/app/v2/notifications/test_post_notifications.py [hunk 1]
```diff
@@ -433,3 +433,17 @@ def test_post_notification_raises_bad_request_if_service_not_invited_to_schedule
...
+
+
+def test_post_notification_raises_bad_request_if_not_valid_notification_type(client, sample_service):
+    auth_header = create_authorization_header(service_id=sample_service.id)
+    response = client.post(
+        '/v2/notifications/foo',
+        data='{}',
+        headers=[('Content-Type', 'application/json'), auth_header]
+    )
+    assert response.status_code == 400
```
Comment (leohemsted) [inline tests/app/v2/notifications/test_post_notifications.py]: or should this be 404

File: app/v2/notifications/notification_schemas.py [hunk 2]
```diff
@@ -192,40 +192,89 @@
...
+    return noti
+
+
+def create_post_letter_response_from_notification(notification, content, subject, url_root, scheduled_for):
+    noti = __create_notification_response(notification, url_root, scheduled_for)
+    noti['content'] = {
+        "body": content,
+        "subject": subject
+    }
+    return noti
```
Comment (imdadahad) [inline app/v2/notifications/notification_schemas.py]: Would be nice to put these functions somewhere else, away from where we define schemas?

File: app/v2/notifications/post_notifications.py [hunk 3]
```diff
@@ -11,28 +13,33 @@
...
 def post_notification(notification_type):
     if notification_type == EMAIL_TYPE:
         form = validate(request.get_json(), post_email_request)
-    else:
+    elif notification_type == SMS_TYPE:
         form = validate(request.get_json(), post_sms_request)
+    elif notification_type == LETTER_TYPE:
+        form = validate(request.get_json(), post_letter_request)
+    else:
+        raise BadRequestError(message='Unknown notification type {}'.format(notification_type))
```
Comment (imdadahad) [inline app/v2/notifications/post_notifications.py]: 👍

File: tests/app/v2/notifications/test_post_letter_notifications.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    template = create_template(service, template_type=LETTER_TYPE)
+
+    data = {
+        'template_id': str(template.id),
+        'personalisation': {'address_line_1': '', 'postcode': ''}
+    }
+
+    error_json = letter_request(client, data, service_id=service.id, _expected_status=400)
+    assert error_json['status_code'] == 400
+    assert error_json['errors'] == [
```
Comment (imdadahad) [inline tests/app/v2/notifications/test_post_letter_notifications.py]: Permissions error returning 400?

File: tests/app/v2/notifications/test_post_notifications.py (hunk 1 above)
Comment (imdadahad) [inline tests/app/v2/notifications/test_post_notifications.py]: Yeah that's probably better, 400 more about body of requests (validating contents) but 404 sounds about right for an non-existent noti type.

File: tests/app/v2/notifications/test_post_letter_notifications.py (hunk 4 above)
Comment (leohemsted) [inline tests/app/v2/notifications/test_post_letter_notifications.py]: We don't currently return 403 for service permissions. See:

https://github.com/alphagov/notifications-api/blob/master/tests/app/v2/notifications/test_post_notifications.py#L293
https://github.com/alphagov/notifications-api/blob/master/tests/app/v2/notifications/test_post_notifications.py#L318
https://github.com/alphagov/notifications-api/blob/master/tests/app/v2/notifications/test_post_notifications.py#L420

**Pull request merged**